pyomo/contrib/iis/mis.py

```diff
@@ -0,0 +1,373 @@
+):
+    """
+    This function attempts to determine why a given model is infeasible. It deploys
+    two main algorithms:
+
+    1. Successfully relaxes the constraints of the problem, and reports to the user
+       some sets of constraints and variable bounds, which when relaxed, creates a
+       feasible model.
+    2. Uses the information collected from (1) to attempt to compute a Minimal
+       Infeasible System (MIS), which is a set of constraints and variable bounds
+       which appear to be in conflict with each other. It is minimal in the sense
+       that removing any single constraint or variable bound would result in a
+       feasible subsystem.
+
+    Args
+    ----
+        model: A pyomo block
+        solver: A pyomo solver object or a string for SolverFactory
+        tee (optional):  Display intermediate solves conducted (False)
+        tolerance (optional): The feasibility tolerance to use when declaring a
+            constraint feasible (1e-08)
+        logger:logging.Logger
+            A logger for messages. Uses pyomo.contrib.mis logger by default.
+
+    """
+
+    # hold the original harmless
+    modified_model = model.clone()
+
+    if solver is None:
+        raise ValueError("A solver must be supplied")
+    elif isinstance(solver, str):
+        solver = pyo.SolverFactory(solver)
+    else:
+        # assume we have a solver
+        assert solver.available()
+
+    # first, cache the values we get
+    _value_cache = ComponentMap()
+    for v in model.component_data_objects(pyo.Var, descend_into=True):
+        _value_cache[v] = v.value
+
+    # finding proper reference
+    if model.parent_block() is None:
+        common_name = ""
+    else:
+        common_name = model.name + "."
+
+    _modified_model_var_to_original_model_var = ComponentMap()
+    _modified_model_value_cache = ComponentMap()
+
+    for v in model.component_data_objects(pyo.Var, descend_into=True):
+        modified_model_var = modified_model.find_component(v.name[len(common_name) :])
+
+        _modified_model_var_to_original_model_var[modified_model_var] = v
+        _modified_model_value_cache[modified_model_var] = _value_cache[v]
+        modified_model_var.set_value(_value_cache[v], skip_validation=True)
+
+    # TODO: For WT / IDAES models, we should probably be more
+    #       selective in *what* we elasticize. E.g., it probably
+    #       does not make sense to elasticize property calculations
+    #       and maybe certain other equality constraints calculating
+    #       values. Maybe we shouldn't elasticize *any* equality
+    #       constraints.
+    #       For example, elasticizing the calculation of mass fraction
+    #       makes absolutely no sense and will just be noise for the
+    #       modeler to sift through. We could try to sort the constraints
+    #       such that we look for those with linear coefficients `1` on
+    #       some term and leave those be.
```

Comment:
Alternatively, we could apply this tool to a version of the model that has as many as possible of these constraints "projected out".